pkg/controller/controller_instance.go

```diff
@@ -70,6 +70,113 @@ func (c *controller) instanceUpdate(oldObj, newObj interface{}) {
 	c.instanceAdd(newObj)
 }
 
+// reconcileInstanceDelete is responsible for handling any instance whose deletion timestamp is set.
+func (c *controller) reconcileInstanceDelete(instance *v1alpha1.Instance) error {
+	// nothing to do...
+	if instance.DeletionTimestamp == nil {
+		return nil
+	}
+
+	finalizerToken := v1alpha1.FinalizerServiceCatalog
+	finalizers := sets.NewString(instance.Finalizers...)
+	if !finalizers.Has(finalizerToken) {
+		return nil
+	}
+
+	// if there is no op in progress, and the instance was never provisioned, we can just delete.
+	// this can happen if the service class name referenced never existed.
+	if !instance.Status.AsyncOpInProgress && instance.Status.Checksum == nil {
+		finalizers.Delete(finalizerToken)
+		// Clear the finalizer
+		return c.updateInstanceFinalizers(instance, finalizers.List())
+	}
+
+	// All updates not having a DeletingTimestamp will have been handled above
+	// and returned early. If we reach this point, we're dealing with an update
+	// that's actually a soft delete-- i.e. we have some finalization to do.
+	// Since the potential exists for an instance to have multiple finalizers and
+	// since those most be cleared in order, we proceed with the soft delete
+	// only if it's "our turn--" i.e. only if the finalizer we care about is at
+	// the head of the finalizers list.
+	serviceClass, servicePlan, brokerName, brokerClient, err := c.getServiceClassPlanAndBroker(instance)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Finalizing Instance %v/%v", instance.Namespace, instance.Name)
+
+	request := &brokerapi.DeleteServiceInstanceRequest{
+		ServiceID:         serviceClass.ExternalID,
+		PlanID:            servicePlan.ExternalID,
+		AcceptsIncomplete: true,
+	}
+
+	glog.V(4).Infof("Deprovisioning Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
+	response, respCode, err := brokerClient.DeleteServiceInstance(instance.Spec.ExternalID, request)
+
+	if err != nil {
+		s := fmt.Sprintf(
+			"Error deprovisioning Instance \"%s/%s\" of ServiceClass %q at Broker %q with status code %d: %s",
+			instance.Namespace,
+			instance.Name,
+			serviceClass.Name,
+			brokerName,
+			respCode,
+			err,
+		)
+		glog.Warning(s)
+		c.updateInstanceCondition(
+			instance,
+			v1alpha1.InstanceConditionReady,
+			v1alpha1.ConditionUnknown,
+			errorDeprovisionCalledReason,
+			"Deprovision call failed. "+s)
```

Comment:
Just a note, nothing necessary to do for this PR - this is the code that we'll need to change in order to fix https://github.com/kubernetes-incubator/service-catalog/issues/908